odh-model-controller/controllers/storageconfig_controller.go

```diff
@@ -0,0 +1,214 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"reflect"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	storageSecretName = "storage-config"
+)
+
+type StorageSecretReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+func newStorageSecret(dataConnectionSecretsList *corev1.SecretList) *corev1.Secret {
+	desiredSecret := &corev1.Secret{}
+	desiredSecret.Data = map[string][]byte{}
+	dataConnectionElement := map[string]string{}
+	storageByteData := map[string][]byte{}
+	for _, secret := range dataConnectionSecretsList.Items {
+		dataConnectionElement["type"] = secret.Annotations["opendatahub.io/connection-type"]
+		dataConnectionElement["access_key_id"] = string(secret.Data["AWS_ACCESS_KEY_ID"])
+		dataConnectionElement["secret_access_key"] = string(secret.Data["AWS_SECRET_ACCESS_KEY"])
+		dataConnectionElement["endpoint_url"] = string(secret.Data["AWS_S3_ENDPOINT"])
+		dataConnectionElement["default_bucket"] = string(secret.Data["AWS_S3_BUCKET"])
+		dataConnectionElement["region"] = string(secret.Data["AWS_DEFAULT_REGION"])
+		jsonBytes, _ := json.Marshal(dataConnectionElement)
+		storageByteData[secret.Name] = jsonBytes
+	}
+	desiredSecret.Data = storageByteData
+	return desiredSecret
+}
+
+// CompareStorageSecrets checks if two secrets are equal, if not return false
+func CompareStorageSecrets(s1 corev1.Secret, s2 corev1.Secret) bool {
+	return reflect.DeepEqual(s1.ObjectMeta.Labels, s2.ObjectMeta.Labels) && reflect.DeepEqual(s1.Data, s2.Data)
+}
+
+// Reconcile
+func (r *StorageSecretReconciler) reconcileSecret(secret *corev1.Secret,
+	ctx context.Context, newStorageSecret func(dataConnectionSecretsList *corev1.SecretList) *corev1.Secret) error {
+	// Initialize logger format
+	log := r.Log.WithValues("secret", secret.Name, "namespace", secret.Namespace)
+
+	dataConnectionSecretsList := &corev1.SecretList{}
```

Comment:
because this is a large procedure, could we add some comments on each logical step this function performs between each block?